src/reference.rs

```diff
@@ -179,6 +179,19 @@ impl<'repo> Reference<'repo> {
         }
     }
 
+    /// Conditionally create a new reference with the same name as the given
+    /// reference but a different OID target. The reference must be a direct
+    /// reference, otherwise this will fail.
+    pub fn set_target(&mut self, id: Oid, msg: &str) -> Result<Reference<'repo>, Error> {
```

Comment:
IMHO I'd be inclined to leave it as `msg: &str`.  If people want an empty log message they can/should pass in empty string.